### PR TITLE
Freebuild Editor

### LINKED ISSUE #240: Freebuild editor
Background:
Editors have two uses in Thrive:
1. When accessed from a particular playthrough, an editor will edit some feature that forms part of that playthrough's state (eg, the structure of your microbe in Microbe stage, or the design of a weapon in Industrial), and thus the data pertaining to the modified feature has to be sent back and incorporated into that playthrough's state.
2. When accessed from the main menu, an editor should provide the user with the ability to make anything they want -- create fantastical creatures, epic (and very expensive) spaceships, etc. These creations, obviously, can't simply be loaded into a regular game.

To do this, we need a bunch of changes:
- Our save-game semantics change up a bit -- "creations" (eg, .*.microbe) are the things that can be looked at and fiddled with in the main-menu-accessed editors. So, if say you wanted to turn your creation into a piece of abstract art, you could load it in the main-menu-accessed editor, and twiddle to your heart's content.

In MainMenu->Editor:
- No mutation point accounting -- or, we instead track the point cost of the microbe. Understandably, the first is much easier than the second
- Every time you start it, it contains the default starter microbe, not the current microbe of the current game -- and of course, it can't change the current microbe either
- All organelles available

In Stage->Editor:
- Cannot load microbes

This issue probably wants to wait until the Microbe Editor <-> Stage data-passing has been completely redone.

### LINKED ISSUE #761: The player should be the only initial species
- [x] player is the only initial species
- [x] doesn't apply to freebuild
- [x] species need to sometimes split off also from the player

### PR DESCRIPTION
**Pull request by 1n48yg**
fixes #240 
fixes #761

### CODE REVIEW

**Comment by hhyyrylainen** [comment]
add to whitelist

**Comment by hhyyrylainen** [comment]
The reason CI is failing on this is, because, I had messed up before, and now only the bsf branch based ones work. This should be rebased on that branch.

File: scripts/gui/main_menu.mjs [hunk 1]
```diff
@@ -227,6 +246,19 @@ function newGame(){
...
 
+function enterFreebuildEditor(){
+    if(common.isInEngine()){
+        Thrive.start();
+        switchToMicrobeHUD();
+        jams.Pause();
+        $("#toolsMenu").slideUp("fast", () => {
+            $("#mainMenu").slideDown("fast");
+        });
+        document.getElementById("topLevelMenuContainer").style.display = "none";
```
Comment (hhyyrylainen) [inline scripts/gui/main_menu.mjs]: Why is this here? the normal route doesn't call this?

File: scripts/gui/main_menu.mjs [hunk 2]
```diff
@@ -227,6 +246,19 @@ function newGame(){
     }
 }
 
+function enterFreebuildEditor(){
+    if(common.isInEngine()){
+        Thrive.start();
+        switchToMicrobeHUD();
```
Comment (hhyyrylainen) [inline scripts/gui/main_menu.mjs]: This should be outside the isInEngine check to work with in-browser preview

File: scripts/gui/microbe_editor.mjs [hunk 3]
```diff
@@ -354,6 +354,7 @@ function onResumeClickedEditor(){
 
 function onExitToMenuClickedEditor(){
     if(common.isInEngine()){
+        document.getElementById("pauseOverlayEditor").style.display = "none";
```
Comment (hhyyrylainen) [inline scripts/gui/microbe_editor.mjs]: Why is this needed? Is this related to the freebuild editor or a general fix?

File: scripts/microbe_stage/microbe_editor/microbe_editor.as [hunk 4]
```diff
@@ -10,13 +11,18 @@
 ////////////////////////////////////////////////////////////////////////////////
 */
 
+bool freeBuilding = false;
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe_editor/microbe_editor.as]: This should be next to the other variables, if this is kept here. However I think that something like "has done freebuild" should be placed in the PlayerData class and the editor should check that when entering. Also that should disable winning. The cleanest way would probably to use PlayerData::setBool and the isBoolSet methods (they allow us to have a bunch of flags related to the current game without adding new methods all the time, and saving them is probably easier).

File: scripts/microbe_stage/microbe_editor/setup.as [hunk 5]
```diff
@@ -1,7 +1,20 @@
...
+    enterEditor(world);
+}
+
+void onFreebuildEditorEntry(MicrobeEditorWorld@ world)
+{
+    freeBuilding = true;
+    enterEditor(world);
+}
+
+void enterEditor(MicrobeEditorWorld@ world)
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe_editor/setup.as]: With my earlier comment needing these separate entry functions goes away.

File: src/ThriveGame.cpp [hunk 6]
```diff
@@ -556,17 +556,28 @@ void
     //     m_impl->m_MicrobeEditorScripts, "microbe editor scripts not loaded");
 
     LOG_INFO("Calling editor setup script onEditorEntry");
+    if(freeBuilding) {
+        ScriptRunningSetup setup("onFreebuildEditorEntry");
```
Comment (hhyyrylainen) [inline src/ThriveGame.cpp]: This is where the PlayerData could be modified if freeBuilding is true.

File: scripts/gui/microbe_editor.mjs (hunk 3 above)
Comment (1n48yg) [reply scripts/gui/microbe_editor.mjs]: Oh this is just a general fix.

File: scripts/gui/main_menu.mjs (hunk 1 above)
Comment (1n48yg) [reply scripts/gui/main_menu.mjs]: Lines 254 to 256 is to reset the menu back to the main page when returning back to it from in game. If I remove the code the game will return to the tools submenu. Although I probably don't need the animation here.

You're right about line 257, I'll remove it.
